pkg/agent/clean/node.go

```diff
@@ -0,0 +1,349 @@
+}
+
+func stopContainers(ctx context.Context, c *client.Client) error {
+	containers, err := c.ContainerList(ctx, types.ContainerListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, container := range containers {
+		isCleanup := false
+		for _, n := range container.Names {
+			if strings.HasPrefix(n, "/"+ContainerName) {
+				isCleanup = true
+				break
+			}
+		}
+
+		if isCleanup {
+			continue // don't stop the cleanup container!
+		}
+
+		config, err := c.ContainerInspect(ctx, container.ID)
+		if err != nil {
+			return err
+		}
+		if strings.HasPrefix(config.Config.Image, "rancher/") {
+			if err := c.ContainerKill(ctx, config.ID, "SIGKILL"); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func waitForK8sPods(ctx context.Context, c *client.Client) error {
+	// wait for up to 5min for k8s pods to be dropped
+	for i := 0; i < 30; i++ {
+		logrus.Infof("checking for pods %d out of 30 times", i)
+		containerList, err := c.ContainerList(ctx, types.ContainerListOptions{})
+		if err != nil {
+			return err
+		}
+
+		hasPods := false
+		for _, c := range containerList {
+			for _, n := range c.Names {
+				if strings.HasPrefix(n, "/k8s_") {
+					hasPods = true
+					continue
+				}
+			}
+			if hasPods {
+				continue //break out if you already found one
+			}
+		}
+
+		if hasPods {
+			logrus.Info("pods found, waiting 10s and trying again")
+			time.Sleep(10 * time.Second)
+			continue
+		}
+
+		logrus.Info("all pods cleaned, continuing on to more rke cleanup")
+		return nil
+	}
+
+	return nil
+}
+
+func Firewall() error {
+	// save docker rules
+	if _, err := exec.Command("sh", "-c",
+		"iptables-save -t filter | egrep -w \"COMMIT|DOCKER|docker0|\\*filter\" > filter.txt").Output(); err != nil {
+		return err
+	}
+
+	if _, err := exec.Command("sh", "-c",
+		"iptables-save -t nat | egrep -w \"COMMIT|DOCKER|docker0|\\*nat\" > nat.txt").Output(); err != nil {
+		return err
+	}
+
+	ipt, err := iptables.New()
+	if err != nil {
+		return err
+	}
+
+	for _, table := range []string{"nat", "mangle"} {
+		logrus.Infof("clearing and deleting iptables table: %s", table)
+		if err := ipt.ClearAndDeleteChain(table, ""); err != nil {
+			return err
+		}
+	}
+
+	if err := ipt.ClearAll(); err != nil {
+		return err
+	}
+	if err := ipt.DeleteAll(); err != nil {
+		return err
+	}
+
+	if _, err := exec.Command("sh", "-c",
+		"iptables-restore filter.txt").Output(); err != nil {
+		return fmt.Errorf("error putting back filter table: %s", err)
+	}
+
+	if _, err := exec.Command("sh", "-c",
+		"iptables-restore nat.txt").Output(); err != nil {
+		return fmt.Errorf("error putting back nat table: %s", err)
+	}
+
+	return nil
+}
+
+func getPaths() []string {
+	pathPrefix := getPrefixPath()
+
+	// all paths on the hosts to rm -rf
+	return []string{
+		hosts.ToCleanCNIConf, hosts.ToCleanCNIBin,
+		"/opt/rke", "/run/secrets/kubernetes.io",
+		"/run/calico", "/run/flannel", "/var/log/containers", "/var/log/pods",
+		hosts.ToCleanCalicoRun,
```

Comment:
why don't these have pathPrefix prefixed on them? Don't we place these in the path prefix as well?

Reply:
you'd have to read rke for history on that.. this was copy/pasted from existing rke functions and then added some too... that's where RKE does it that's where we have to clean them up from.